nimble-wafer-map component spec document (#805)
ni/nimble: packages/nimble-components/src/wafer-map/specs/README.md

* Allowing to set the orientation of the wafer
* Allowing an option to select specific color schemes for the wafer
* Allowing to zoom in/out on the wafer-map widget
* Allowing panning whilst the wafer is zoomed
* Dynamic size adaptability
* Scalable for very large number of data points

### Risks and Challenges

The biggest risk and challenge are regarding the rendering of the component when large numbers of data points are present. 
This widget can and will be used to represent the results of a single wafer and a summary of multiple wafers (data which represents also known as a LOT) in order to recognize patterns in failing sections.
A LOT contains data about multiple wafers, and one of the functionalities of our main feature is to split up a LOT (and display it) into individual wafers.
This means that we might be displaying several `nimble-wafer-map` components on a single page each containing several hundred or thousands of data points, each having the individual functionality mentioned in the "Features" section of this spec document.
We are planning to perform initial benchmarking on this aspect, but not planning to optimize it in the first release.

> ### Prior Art/Examples

Reviewer: nit: formatting
```suggestion

### Prior Art/Examples
```

Please see below a few screenshots regarding the current `wafer-map` component:

> ![WaferMap Example](file:///Resources/wafer_map_example.png)

Reviewer 2: These urls should be relative path urls in order to render on github, ie:

```suggestion
![WaferMap Example](./Resources/wafer_map_example.png)
```

You may have to update other image urls as well

While zooming in on the component, the user should also be able to pan (drag) the wafer map by holding the left mouse button and moving the cursor on the XY axis.

From a developer perspective, it's very important that the `nimble-wafer-map` component can be integrated with the MicroStrategy Custom plugin framework as this framework provides functionalities as follows:
* Tooltip when hovering any if the datapoints withing the wafer map
* Wafer map legend
* MicroStrategy specific context menus on right-click action

Information regarding the basic MicrosStrategy **Custom Plugin Workflow** can be found [here](https://www2.microstrategy.com/producthelp/Current/VisSDK/Content/topics/HTML5/Understanding_the_workflow.htm)
Also, the visualisation widget has no active controls, so there is no way to alter the visual aspect or data post rendering. Every input parameter must be provided through the constructor of the wafer map component.

### API

*The key elements of the component's public API surface:*
* Component Name: `nimble-wafer-map`
* Props/Attrs: 
  * `die[]` - this represents the input data, which fills the wafer map with content.\

>   The **die**  object contains the following attributes:

Reviewer: Naming suggestion: `x`, `y`, and `value`. Follows nimble's camel case convention and is a bit less verbose.

	* XCoorrdianate: integer
	* YCoordinate: integer
	* Value: float
  * `quadrant` - represents the orientation fo the dies on the wafer map (the layout of the values on the dies). It can be represented by an Enum with the following values: 

> 	* TopLeft - ![Top Left Quadrant](file:///Resources/top_left.png)

Reviewer 2: Nimble directory names and files generally use lower kebab case formatting (besides special cases such as `README.md`), can you update the directory and file names to match the repo?

ie:
```suggestion
	* TopLeft - ![Top Left Quadrant](./resources/top-left.png)
```

Author: Fixed as suggested.

	* BottomLeft - ![Bottom Left Quadrant](file:///Resources/bottom_left.png)
	* TopRight - ![Top Right Quadrant](file:///Resources/top_right.png)
	* BottomRight - ![Bottom Right Quadrant](file:///Resources/bottom_right.png)
  * `orientation` - represent the orientation of the notch on the wafer map outline (only visual). As only four static orientations are possible, it can be represented by an Enum with the following values: top, bottom, left, right.

>   * `color_scale` - represents the color spectrum which shows the status of the dies on the wafer.

Reviewer: Could you describe the type you expect for color scale? Array of strings of hex colors? Will we support [all the ways that CSS specifies color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) like named colors (`"red"`)?

Also note that Nimble's convention is to use `camelCase` for property names so I'd propose `colorScale`.

Reviewer: Thanks for clarifying the plan. I tentatively think I'd advocate not to expose d3's color types as part of the public API and rather see if we can find a standard way to represent color that can be converted to d3's types. That way we're not tied to d3 if we decide to change our underlying implementation.

>   * `number_of_characters` - represents the number of characters allowed to be displayed within a single die. As the die values are represented by Floating point numbers, we must have the liberty of limiting how many characters we are willing to display within a single die.

Reviewer: I'd propose a more specific name that explains which text this is controlling. Maybe `dieCharacterCount`? `maxDieLabelLength`?

>  Please note that all these attributes are only accessible through the constructor of the component. Once set, the only way to modify them is to re-render the component.

Reviewer: Just to think through some of the consequences of this:
1. there is no attribute API for this component (e.g. I can't set `<nimble-wafer-map orientation='right'>`
2. the property API is read-only (e.g. in JS code I can read `waferMap.orientation` but not write it)
3. there is no way to update the data in the wafer map other than constructing a new wafer map (I can't set `waferMap.dies = [...];` or call a method like `refresh()`/`render()`

If my understanding is correct then I'd propose rewriting like this:

```suggestion
 Please note that all these properties are only settable through the constructor of the component, not via properties or attributes. Once set, the only way to modify them is to initialize a new component.
```

Reviewer 3: I'd like to dig into this a bit more from a reasoning perspective. Why shouldn't the API be settable? What makes a wafer-map distinct enough in its usage, from other data visualization components like graphs, that suggest that a user should not be able to modify its state after construction? At the moment, being somewhat naive, I don't see the benefit.

Reviewer 2: > Please note that all these attributes are only accessible through the constructor of the component. Once set, the only way to modify them is to re-render the component.

Custom element APIs are a bit different in that you cannot pass properties to a constructor for an element. The browser owns the constructor and calls it internally, we can't pass options to it.

For example, if I have the HTML in a page of `<input type="button">` that is equivalent to the code:

```ts
const el = document.createElement('input');
input.type = 'button';
```

So behaving like a DOM element is a **requirement** for a custom element based API. Configuration has to be via writeable properties and attribute.

**This should not be a major concern.** The FAST library we build on gives us [internal observability behaviors](https://www.fast.design/docs/fast-element/observables-and-state/#internal-observation) to make updates straightforward.

Roughly the approach would be to create attributes, listen for attribute changes for properties that require recreating re-rendering, and re-rendering the internal state on a `requestAnimationFrame`.

Some pseduocode:
```ts
class WaferMap extends FastElement {
    private renderQueued = false;

    @attr
    orientation;

    @attr
    quadrant;

    orientationChanged() {
       this.queueRender()
    }

   quadrantChanged() {
       this.queueRender()
   }

   private queueRender() {
      if (!this.renderQueued) {
          this.renderQueued = true;
          window.requestAnimationFrame(() => {this.render()})
      }
   }
   
   private render () {
        this.renderQueued = false;
        // Re-render the component
   }
}

```

So some insights are because we would render on requestAnimationFrame this gives us some benefits:

- We only do work for re-rendering the element as the browser draws frames on the screen
- If the user rapidly changes settings we do not waste time doing unnecessary work before the component is rendered again to draw on the screen

So the spec language I would use would be something like:

```suggestion
    As the wafer map will use the d3 library to draw it's contents, attribute value changes will cause the component to re-render on updates. To avoid unnecessary re-rendering from rapid attribute updates, the component will re-render its content during `requestAnimationFrame`.
```

Reviewer: Thanks @rajsite. I had been dissuading the team from exposing an attribute API because so many of the properties are complex data types that would have to be encoded/decoded to (presumably JSON) strings. Do you think that's a compelling argument to only expose a property API or is there still value in exposing an attribute API even if it requires some slightly clunky data conversion? I don't feel strongly so could be easily swayed.

@atmgrifter00's question about exposing a settable API is probably moot given the inability to set data in the constructor. But I'm still interested in the original rationale. Is it just that current use cases don't require updating the wafer map after creation or is there a technical challenge to doing so.

Reviewer 2: @jattasNI @zszilagy 

I'd advocate that:
 - the simple properties (numbers, boolean, enums, strings) should be configurable via properties and attributes
   - quadrant, orientation, dieCharacterCount, disabled
 - the complex properties should only be configurable via properties and have no attribute
   - die, and colorScale

And based on the recent push I would advocate we keep the following methods:
- focus()

But remove the following methods:
- setWaferMapOrientation(): orientation should be a writeable property and attribute as mentioned above
-  setWaferMapQuadrant(): quadrant should be a writeable property as mentioned above
- render() should not be a method, the above properties, attributes, and methods should queue rendering as needed

Also @zszilagy you may need to click the `load more` button if you intended to respond to this thread:

![image](https://user-images.githubusercontent.com/1588923/201205640-58642261-7411-46dd-90a7-cf63b035019a.png)

Author: @rajsite Thanks for the suggestions. I included them into the spec and will resolve this thread once I fixed everything.

</template>
```
The first SVG element with the "zoomContainer" class attribute contains two circles and a rectangle.
* The first circle represents the outline of the wafer map component.
* The second circle, together with the rectangle is part of a solution which creates the illusion that there is a notch present on the wafer map frame.

The second SVG element with the "WaferMapArea" class attribute represents the area where the dies get rendered.
The rectangle present here is used as the "highlight" which allows us to identify individual dies in the canvas.

- Slot Names: (none)
- Host Classes: (none)
- Slotted Content/Slotted Classes: (none)
- CSS Parts: (none)

### Angular integration

The `nimble-wafer-map` component does not take user input directly from the webpage where it's embedded. All the inputs must be provided through the constructor of the component so in the current case there is no point using ControlValueAccessors.

### Blazor integration

Blazor support is out of scope for the current implementation.

### Visual Appearance

For the initial component, we are going with the original wafer map appearance. 
![WaferMap Example](file:///Resources/wafer_map_example.png)

> 

Reviewer: I think we're missing a section on Accessibility. Long term, we'd want to discuss questions like:
1. how is the wafer map represented to a screen reader? The WAI guideline on [complex images](https://www.w3.org/WAI/tutorials/images/complex/) might apply; it suggests providing high-level alt text describing the graph
2. how does the user navigate the data with a keyboard? Can you tab into the content and move around to different dies with the arrow keys?

Even if we decide not to implement an accessible component initially, the spec gives us a good opportunity to record that decision and lay out a vision for what we'd achieve eventually.

* As an initial step, we submit a skeleton of the `nimble-wafer-map` component into the main branch. This will serve as a starting point on which the complete functionality will be built on.
* Once the skeleton is in place, we start to work on and submit individual pieces of the functionality so it's not very overwhelming to review the code.

### States

We are not planning to allow the user to generate an “empty" wafer map, so we must ensure that data is provided when instantiating a new `nimble-wafer-map` component.
The only action which we could consider as a "state transition" is the zoom functionality which was specified in the sections above.
We expect that the end user be able to zoom in on the dies withing the wafer map and be able to zoom out by using the mouse wheel.
Also, the user must be allowed to pan the zoomed in wafer map into any direction of its XY axis by holding the left mouse button and moving the mouse cursor around.

Another interaction, which must be part of the component is the ability to identify the location of the mouse pointer on the wafer map die area.

### Globalization

> Currently there will be no string data presented by this component. Might change in the future, but for now it will contain only numerical data.

Reviewer: ```suggestion
Currently there will be no string data presented by this component. Might change in the future, but for now it will contain only numerical data. Similar to the `nimble-number-field`, initially we don't plan to support locale-specific numeric formatting.
```

### Security

N/A

### Performance

As mentioned in the "Risks" section, the biggest challenge from the performance standpoint is the representation of large number of datapoints and interacting with these datapoints in the browser.
Currently the only thing we would like to achieve is to measure (benchmark) the time it takes a `nimble-wafer-map` component to be loaded with various number of datapoints.

### Dependencies

> 

Reviewer: It would be good to confirm that the new dependency on `d3` doesn't bloat other clients. I would hope that:
1. clients that don't use the wafer map will not pay any bundle size or load time cost (It's ok if installing Nimble causes d3 to be installed, but don't want it to impact built apps)
2. clients that do use the wafer map won't have to download the entire library if we only use a small part of it

If we consume d3 using the ES `import` statement then this should all work for free.

Author: We import the following modules from d3 library:
- scaleBand
- ScaleBand
- ScaleLinear
- scaleLinear
- ScaleOrdinal
- scaleOrdinal
- select
- Selection
- zoom
- ZoomBehavior
- zoomIdentity
- zoomTransform
- ZoomTransform

Author: Currently we have the whole d3 library as dependency, but basically what we need is d3-scale, d3-zoom and d3-selection. 
I tested our current component and it builds and behaves as expected even using only these 3 as dependency instead of complete d3. 
The only downside is that the versions of these components are not aligned all the time: in d3 v:7.6.1 the following versions are used:
d3-zoom: v3
d3-scale: v4
d3-selection: v3

It's a bit more annoying to keep up to date 3 packages instead of the d3 library, however, this way we only use 3 out the complete 30 which comes with d3.

Reviewer 2: Because it is just a few packages I think depending on them individually is the way to go. It will keep bundles small without relying on dead-code elimination features and I don't think will be a significant maintenance burden.

> If we consume d3 using the ES import statement then this should all work for free.

If we did rely on dead-code elimination and smart bundlers then this statement is true as long as we don't do builk imports like `import * as d3 from 'd3';`. We would have to be rigorous about importing only exactly what is needed, ie `import {zoom} from 'd3';` and rely on bundlers to dead-code eliminate the unused d3 exports.

Because we have relatively few packages that we are depending on from d3 I would vote for just using the separate packages but I'm open to either approach.

Reviewer: Thanks, that's really useful background. It looks like d3 is using semantic versioning so the version mismatch might not be a problem; that just means that there have been more revisions of the `d3-scale` package than the others, not necessarily that they can't work together. Hopefully npm can just resolve any compatible differences in their sub-dependencies and show us an error if they are incompatible.

So I'd vote to depend on just the smaller packages, not all of d3. That reduces the exposure to security vulnerabilities and package bloat. I agree it's a bit of extra work to keep them up to date but long term we can automate that so it shouldn't be much of a burden.

And regarding my original point about eliminating bundle cost of d3 for clients not using the wafer map, I think the fact that you `import` d3 classes (rather than including a d3.js script on the page for example) should solve that issue.

Reviewer: Looks like Milan and I replied at the same time. Thankfully we gave the same advice 😮‍💨

The only 3rd party dependency on which the current wafer map component depends on is the [D3(Data-Driven Documents)](https://d3js.org/) library.
We use parts of this library in our zooming and rendering functionality.

> 

Reviewer: @fredvisser Looks like d3 uses an ISC license. I don't have experience with that license. It looks permissive like MIT but I want to confirm: is it generally allowed in our software?
https://github.com/d3/d3/blob/main/LICENSE

Reviewer 4: Yep - ISC is 👍

> 

Reviewer: Do you have a sense for our plan for updating the version of d3 going forward?
1. how often do they publish updates? Are they usually pretty stable or lots of breaking changes?
2. how often do they have security vulnerabilities (directly or in their sub-dependencies) and how quickly do they fix them?

I ask because we don't have automated dependency updates turned on for this repo yet because we have so few third-party dependencies. If you think it would be valuable to update d3 frequently, we could prioritize the work to update it automatically.

Author: We use pretty much the basic functionality from the d3 library. 
I verified they repo and the last breaking change happened with the release of 7.0.0 release. (June 2021).
From the release history the release pattern is the following:
1. Major version change seems to happen once a year
2. Minor version change happens ~every three months
3. Revision changes are random - sometimes revision changes in every few days, weeks

We don't expect to be aligned with the latest version all time, however, we expect not to miss important fixes regarding modules we use or security updates.

Reviewer: Cool. We have basic security auditing on our pipeline so we should get notified of vulnerabilities and that will trigger us to look for security updates.

Longer term as part of #291 I would like to turn on Renovate or Dependabot in this repo to automatically create PRs to update deps on some cadence. But given the versioning you described I don't think it will be too burdensome to keep our versions pinned most of the time and to update them manually when needed.

### Test Plan

The current test plan is that moving forward with the development, we write unit tests which validate the core functionality.
All tests written for this component should follow the standards of the nimble repository. 

> ### Documentation

Reviewer: nit: formatting

```suggestion

### Documentation

```

> We will write documentation on the `nimble-wafer-map` component, focusing on general usage, implementation and API. 

Reviewer: It's good to clarify where the documentation will go.

Regarding implementation documentation, we've mostly documented the high level plan in the form of this spec but not gone into much more detail elsewhere (other than occasional inline comments). Do you think we'll need more docs for this component?

```suggestion
We will write Storybook documentation on the `nimble-wafer-map` component, focusing on general usage and API. 
```

> # Nimble Wafer Map

Reviewer: Heads up that the PR pipeline is failing because of lint errors in this new file. Before you submit you can run `npm run format` from the root of the repo to fix this (after running `npm install` to install the lint tooling of course).

This means that we might be displaying several `nimble-wafer-map` components on a single page each containing several hundred or thousands of data points, each having the individual functionality mentioned in the "Features" section of this spec document.
We are planning to perform initial benchmarking on this aspect, but not planning to optimize it in the first release.

### Prior Art/Examples

Please see below a few screenshots regarding the current `wafer-map` component:

![WaferMap Example](./Resources/wafer_map_example.png)

## Design

From a user’s perspective one of the most important actions which this component should be able to achieve is to zoom in and out on the datapoints. This functionality should occur when the mouse pointer hovers the `nimble-wafer-map` component and a mouse wheel action is performed.
While zooming in on the component, the user should also be able to pan (drag) the wafer map by holding the left mouse button and moving the cursor on the XY axis.

From a developer perspective, it's very important that the `nimble-wafer-map` component can be integrated with the MicroStrategy Custom plugin framework as this framework provides functionalities as follows:

> * Tooltip when hovering any if the datapoints withing the wafer map

Reviewer 3: ```suggestion
* Tooltip when hovering any of the datapoints within the wafer map
```

* Wafer map legend
* MicroStrategy specific context menus on right-click action

> Information regarding the basic MicrosStrategy **Custom Plugin Workflow** can be found [here](https://www2.microstrategy.com/producthelp/Current/VisSDK/Content/topics/HTML5/Understanding_the_workflow.htm)

Reviewer 3: ```suggestion
Information regarding the basic MicroStrategy **Custom Plugin Workflow** can be found [here](https://www2.microstrategy.com/producthelp/Current/VisSDK/Content/topics/HTML5/Understanding_the_workflow.htm)
```

Also, the visualisation widget has no active controls, so there is no way to alter the visual aspect or data post rendering. Every input parameter must be provided through the constructor of the wafer map component.

### API

*The key elements of the component's public API surface:*
* Component Name: `nimble-wafer-map`
* Props/Attrs: 
  * `die[]` - this represents the input data, which fills the wafer map with content.\
  The **die**  object contains the following attributes:
	* x: integer
	* y: integer
	* alue: float
  * `quadrant` - represents the orientation of the dies on the wafer map (the layout of the values on the dies). It can be represented by an Enum with the following values: 
	* TopLeft - ![Top Left Quadrant](./Resources/top_left.png)
	* BottomLeft - ![Bottom Left Quadrant](./Resources/bottom_left.png)
	* TopRight - ![Top Right Quadrant](./Resources/top_right.png)
	* BottomRight - ![Bottom Right Quadrant](./Resources/bottom_right.png)
  * `orientation` - represent the orientation of the notch on the wafer map outline (only visual). As only four static orientations are possible, it can be represented by an Enum with the following values: top, bottom, left, right.

>   * `colorScale` - represents the color spectrum which shows the status of the dies on the wafer.\

Reviewer: Continuing the discussion from [this thread](https://github.com/ni/nimble/pull/805#discussion_r1017187303) in a new one so GitHub associates it with the new location in the file...

Thanks for clarifying the plan. I think I'd advocate not to expose d3's types as part of the public API and rather see if we can find a standard way to represent the data that can be converted to d3's types. That way we're not tied to d3 if we decide to change our underlying implementation.

It looks like the scale ordinal is just a pair of arrays for color and values. Could we just make those be two separate properties?

Reviewer 3: Conceptually I don't like them as two different properties, because they are fundamentally associated with one another. One idea (and I'm just shooting from the hip) is coming up with an interface (e.g., `IColorScale`) that would return a color (likely hex string) for a given numeric value. We could then implement that interface with a custom element, which could then be slotted into the wafer-map component:

```
<wafer-map>
    <d3-color-scale slot="color-scale" colors="red, blue, green" values="1, 2, 3">
</wafer-map>
```

Reviewer: I spoke with the dev team on a call this morning. Having a separate slotted object to represent a scale seems like an interesting long term direction and we should record it in the spec as an alternative. It would be especially useful if we want to include more information in a scale. 

But for the first pass to reduce scope growth we think two arrays is an ok place to start. We can still switch to the scale object in the future, it would just be a breaking change (not a big deal since we'll know about all the clients initially).

![WaferMap Example](./Resources/wafer_map_example.png)

## Implementation

As mentioned above, this is not a component that we start from scratch, but a visualization that already exists which is outdated and not futureproof. 
As a first step, we are working on the "bloat removal" which is basically the removal of unwanted but exiting features (legend, tooltip etc.) and dependencies.

Original component: [wafermap](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=/libraries/wafermap&version=GBfeature/microServicesSplit)\
Once the "bloat removal" is finished, the next step is to adapt the current 

### Code submission strategy

In order to make the process of submitting the re-written wafer map component easier for us and for the code reviewers as well, we propose the following bottom-up approach:
* As an initial step, we submit a skeleton of the `nimble-wafer-map` component into the main branch. This will serve as a starting point on which the complete functionality will be built on.
* Once the skeleton is in place, we start to work on and submit individual pieces of the functionality so it's not very overwhelming to review the code.

> ### States

Reviewer: Another set of states to consider are the common ones used by other components: 
1. `disabled` renders the component dimmed out and turns off interactivity. This seems useful to support.
2. `focus` adds a green box around the component when it has tab focus. This seems useful to support.
3. Most of our components render differently depending on the theme of the page. As you mention, since the client can set specific die colors I don't think we need to change those in response to theme. But the chrome around the wafer map (the circle and notch border) should probably change to be visible on different background colors.
4. Some other common ones are `appearance` for showing different styles like removing the chrome and `read-only` for turning off interactions without making it disabled. I don't think we need either for this component initially.

Author: 1. For the initial release, the only thing this would disabled is the mouse hover over the dies and with it the tooltip. But we can add this for futureproofing. - added to the spec
2. I agree with this. This could be useful when we implement the Accessibility features in the future. - added to the spec.
3. This make sense, however, we would not like to give the liberty for the user to specify the chrome color manually. It would make more sense if we detect the theme used when rendering the wafer map and switch the color of the chrome to something standard set by us. (e.g. NI green when we use a dark theme and black when we use light color). We will discuss this with the team and include a section on this into the Design section.
4. I agree with this. Maybe we can take advantage of this when the wafer map offers a deeper level of interaction.

Reviewer: I agree that autodetecting the theme to set the chrome color would be the right way to go. Note that this might depend on some infrastructure that's needed for #783 too: our color tokens mostly used from CSS right now but you'd need to use them from TS and have a way to be notified when the token value changes because the theme has changed so that you can rerender the border. Should be fairly easy for us to add, just want to mention it to link the work item to this discussion.

Author: @jattasNI are you planning to have this implemented in the near future? I saw that the linked task (#787 ) is expected to be implemented in Q4 this year,

Reviewer 2: > It would make more sense if we detect the theme used when rendering the wafer map and switch the color of the chrome to something standard set by us.

> The first SVG element with the "zoomContainer" class attribute contains two circles and a rectangle.
> -   The first circle represents the outline of the wafer map component.
> -   The second circle, together with the rectangle is part of a solution which creates the illusion that there is a notch present on the wafer map frame.

If the chrome is defined by the inline svg then we can style the svg the css using the existing token infrastructure. We do a similar thing to style the icons which are inline svgs with css, example: 

https://github.com/ni/nimble/blob/197370ec27961996d864f6356e8208c55ae4a99a/packages/nimble-components/src/icon-base/styles.ts#L43-L47

Reviewer: > @jattasNI are you planning to have this implemented in the near future?

From @rajsite's comment above it seems like we likely won't need it but if it turns out we do then we can increase it's priority. Can't commit to it for sure before we know how big the work will be but we want to make you successful.

Author: Sure. If not, we will go with @rajsite's suggestion.

> We are not planning to allow the user to generate an “empty" wafer map, so we must ensure that data is provided when instantiating a new `nimble-wafer-map` component.

Reviewer 3: As suggested in other comments, I think ultimately we _will_ have a concept of an "empty" wafer-map, as we won't be able to insist that data is present before the component is in the DOM. Thus, I think the question is do we need to come up with a design for an empty wafer-map?

Author: Good point. In that specific case we could draw just the outline of the wafer map (the circle with the notch) without the dies in it. The concept of "not having an empty wafer map" I was referring to the fact that we would like not allow the user to create an empty visualization, which will be filled afterwards with data. (for example having an "Add data" button which allows to upload a CSV representing the wafer map data.). That is why I mentioned that it's mandatory to provide the data when instantiating a new wafer map component. But in case the data is not correct and no dies can be drawn into the wafer map area, I agree, that an empty wafer map should be displayed.

The only action which we could consider as a "state transition" is the zoom functionality which was specified in the sections above.

> We expect that the end user be able to zoom in on the dies withing the wafer map and be able to zoom out by using the mouse wheel.

Reviewer 3: ```suggestion
We expect that the end user be able to zoom in on the dies within the wafer map and be able to zoom out by using the mouse wheel.
```

Shadow DOM:

```html
<template id="WaferMapTemplate">
    <div class="WaferMapContainer">
        <svg>
            <g class="zoomContainer">
                <g>
                    <circle></circle>
                    <circle></circle>
                    <rect></rect>
                </g>
            </g>
        </svg>
        <div class="WaferMapArea">
            <canvas />
            <svg>
                <rect></rect>
            </svg>
        </div>
    </div>
</template>
```

The first SVG element with the "zoomContainer" class attribute contains two circles and a rectangle.

-   The first circle represents the outline of the wafer map component.
-   The second circle, together with the rectangle is part of a solution which creates the illusion that there is a notch present on the wafer map frame.

The second SVG element with the "WaferMapArea" class attribute represents the area where the dies get rendered.
The rectangle present here is used as the "highlight" which allows us to identify individual dies in the canvas.

-   Slot Names: (none)
-   Host Classes: (none)
-   Slotted Content/Slotted Classes: (none)
-   CSS Parts: (none)

### Angular integration

> The `nimble-wafer-map` component does not take user input directly from the webpage where it's embedded. All the inputs must be provided through the constructor of the component so in the current case there is no point using ControlValueAccessors.

Reviewer 2: ControlValueAccessors are for Angular Form participation like for buttons, text fields, radio groups, etc. So we will not for a ControlValueAccessor but we will require an Angular directive to expose the typed properties (very straightforward boilerplate).

```suggestion
The `nimble-wafer-map` component will not participate in Angular Forms and does not need a ControlValueAccessor. It will have a standard Angular Directive to expose typed properties to Angular.
```

### Blazor integration

Blazor support is out of scope for the current implementation.

### Visual Appearance

For the initial component, we are going with the original wafer map appearance.
![WaferMap Example](./Resources/wafer_map_example.png)

## Implementation

As mentioned above, this is not a component that we start from scratch, but a visualization that already exists which is outdated and not futureproof.

> As a first step, we are working on the "bloat removal" which is basically the removal of unwanted but exiting features (legend, tooltip etc.) and dependencies.

Reviewer 2: ```suggestion
As a first step, we are working on the "bloat removal" which is basically the removal of unwanted but existing features (legend, tooltip etc.) and dependencies.
```

Original component: [wafermap](https://dev.azure.com/ni/DevCentral/_git/op-web-portal?path=/libraries/wafermap&version=GBfeature/microServicesSplit)\

> Once the "bloat removal" is finished, the next step is to adapt the current

Reviewer 2: Was there more to this sentence? It looks like it is cut-off.

Author: @rajsite I missed this one. I added the missing information.

Another interaction, which must be part of the component is the ability to identify the location of the mouse pointer on the wafer map die area.

Furthermore `disabled` and `focused` state will also be supported by the component.

### Accessibility

Important: this is out of scope for the current implementation.\
Since the `nimble-wafer-map` component was designed to display tens of thousands of data points, it's challenging to provide Accessibility functionality which can read out loud the values displayed in a meaningful manner.
Based on the WAI documentation on [Complex Images](https://www.w3.org/WAI/tutorials/images/complex/), in the future we could provide a paragraph containing information about the name of the wafer that is displayed, the LOT where the wafer comes from and a general overview (e.g. % of the faulty dies on the wafer) of the status. This might give a pretty good description about what the displayed nimble component shows to the user.
Please note that in the initial implementation we do not plan to gather and summarize any of this information so this will be part of future updates to this component.

> Currently the only way to navigate and interact with the wafer map is only by using the mouse. Future updates might include functionality which allows the user to tab into the wafer map area and move around in the die area using the keyboard arrows.

Reviewer 2: We can choose to make the accessibility good in a follow-up but we should do the minimum to make sure we don't break accessibility of existing pages today.

I think we should:
- Make sure we have tab focus states (as already described above)
- Make sure to choose a good role to represent the visual. I think [`role="img"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/img_role) or [`role="presentation"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/presentation_role) are likely good and I think I lean towards `img` role

Author: Thanks @rajsite. We will go with your suggestion on this.

### Globalization

Currently there will be no string data presented by this component. Might change in the future, but for now it will contain only numerical data. Similar to the `nimble-number-field`, initially we don't plan to support locale-specific numeric formatting.

### Security

N/A

### Performance

> As mentioned in the "Risks" section, the biggest challenge from the performance standpoint is the representation of large number of datapoints and interacting with these datapoints in the browser.

Reviewer 2: It sounds like this is important and mentioned a couple times that lots of data may be passed to the component. Do you have expected performance requirements?

I would be interested in:
- Expected data set size max ranges, up to 1K, 10K, 100K?
- Expected update time? Data refreshes every 10s, 1s, 100ms?
- How long after setting to render? Draw is within 10ms, 100ms, 1s, 3s?

Author: For the initial release we do not have any performance expectation from the component. The only requirement for now is to provide empirical data on render time for: single wafer with e.g. 50k data points, multiple wafers with 50k data points, how increasing the number of wafers affects the rendering time etc. We would like to do this benchmarking once we are done with the implementation.

Reviewer 2: > For the initial release we do not have any performance expectation from the component.

I'm sure there are some expectations. If it took 5 seconds to render 10 data points I doubt that would be acceptable. 😋
If we have some basic lower bounds based on expected data set sizes that would be great to capture in the spec.

> The only requirement for now is to provide empirical data on render time for: single wafer with e.g. 50k data points, multiple wafers with 50k data points, how increasing the number of wafers affects the rendering time etc. We would like to do this benchmarking once we are done with the implementation.

If those are the initial benchmarks that sounds worth capturing in the spec and the spec can be updated as we refine the performance goals 👍

Author: @rajsite yeah 10 data points in 10 seconds would be a disaster. I'll set some goals for the benchmarking, something like: 
100 data points
1k data points
10k data points
100k data points
 
just to have a reference point.

Author: Updated the performance requirements with a few estimates. However, these are not based on real world data, they set a pretty clear benchmark for future optimizations.

### Prior Art/Examples

Please see below a few screenshots regarding the current `wafer-map` component:

![WaferMap Example](./Resources/wafer_map_example.png)

## Design

From a user’s perspective one of the most important actions which this component should be able to achieve is to zoom in and out on the datapoints. This functionality should occur when the mouse pointer hovers the `nimble-wafer-map` component and a mouse wheel action is performed.
While zooming in on the component, the user should also be able to pan (drag) the wafer map by holding the left mouse button and moving the cursor on the XY axis.

From a developer perspective, it's very important that the `nimble-wafer-map` component can be integrated with the MicroStrategy Custom plugin framework as this framework provides functionalities as follows:

-   Tooltip when hovering any of the datapoints within the wafer map
-   Wafer map legend
-   MicroStrategy specific context menus on right-click action

Information regarding the basic MicroStrategy **Custom Plugin Workflow** can be found [here](https://www2.microstrategy.com/producthelp/Current/VisSDK/Content/topics/HTML5/Understanding_the_workflow.htm)

> Also, the visualisation widget has no active controls, so there is no way to alter the visual aspect or data post rendering. Every input parameter must be provided through the constructor of the wafer map component.

Reviewer 2: Missed a spot discussing the removed constructor language

```suggestion
Also, the visualisation widget has no active controls, so there is no way to alter the visual aspect or data post rendering.
```

### API

_The key elements of the component's public API surface:_

-   Component Name: `nimble-wafer-map`
-   Props/Attrs:
    -   `die[]` - this represents the input data, which fills the wafer map with content.\
        The **die** object contains the following attributes:
        -   x: integer
        -   y: integer
        -   value: float
    -   `quadrant` - represents the orientation of the dies on the wafer map (the layout of the values on the dies). It can be represented by an Enum with the following values:

>     -   TopLeft - ![Top Left Quadrant](./Resources/top_left.png)

Author: I changed the "enum" word to "const" and added a link about the recommended implementation.

        In the following example the colorScale object is defined as `WafermapColorsScale(['red', 'blue', 'green'], ['1', '2', '8']);`\
        The generated wafer using this color scale is: ![color_scale](./Resources/color_scale.png)
    -   `dieCharacterCount` - represents the number of characters allowed to be displayed within a single die. As the die values are represented by Floating point numbers, we must have the liberty of limiting how many characters we are willing to display within a single die.
    -   disabled - it's represented by a boolean value and reffers to the state of the `nimble-wafer-map` component. If true, the component should be rendered dimmed out and no user interaction should be allowed

The `quadrant`, `orientation`, `dieCharacterCount` and `disabled` properties will be configurable via properties and attributes.
The `die` and `colorScale` properties will be configurable only via properties and will not have attributes.

Methods: The following methods will be exposed in the public API:

> -   disable() - renders the component dimmed out and turns off interactivity.

Reviewer 2: As we have the `disabled` property we no longer need a disable method. Using only a disabled property also aligns better with other HTML components

Author: Removed as suggested.

-   focus() - adds a green box around the component when it has tab focus.

Events: The events mentioned below will all be handled internally by the nimble component and they will not be part of the public API. In the initial implementation the following events should be handled:

-   Zoom in while hovering - this action gets executed whenever the mouse pointer hovers the `nimble-wafer-map` component and a wheel event (scroll up) gets triggered
-   Zoom out while hovering - this action gets executed whenever the mouse pointer hovers the `nimble-wafer-map` component and a wheel event (scroll down) gets triggered
-   Drag while zoomed - this event gets triggered whenever the `nimble-wafer-map` component is zoomed in (larger than 100%) and whilst the left mouse button is held the pointer moves to any direction within the wafer-map canvas
-   Mouse hover - this event gets triggered whenever the mouse pointer is hovering any of the die elements within the wafer map. We only must detect this in the nimble component, proper handling will be done in the MicroStrategy wrapper. (Tooltip triggering)

### Anatomy

Shadow DOM:

```html

> <template id="WaferMapTemplate">

Reviewer 2: nit: in practice we would not set an `id` on the `template` element. That would be reflected outside of the shadowroot which would not be expected behavior (the user sets their own `id` on elements). I don't think this will be a major concern, just wanted to point it out.

Author: I removed the ID so it appears as the expected layout in the spect.
